fala/apps/adviser/forms.py

```diff
@@ -69,67 +70,88 @@ def __init__(self, *args, **kwargs):
         name = data.get("name")
 
-        # Check if both postcode and name are missing
+        # Add error if both postcode and name are missing
         if not postcode and not name:
             self.add_error("postcode", _("Enter a postcode"))
             self.add_error("name", _("Enter an organisation name"))
             return data
 
-        # Validate postcode if provided and region is either England/Wales or Scotland
-        if postcode and self.region in {Region.ENGLAND_OR_WALES, Region.SCOTLAND}:
-            if not self._valid_postcode(postcode):
+        # Do search on organisation name even if postcode missing
+        if not postcode and name:
+            self._region = Region.ENGLAND_OR_WALES
+            return data
+
+        # Validate postcode if provided and then determine region
+        if postcode:
+            valid_postcode = self.validate_postcode_and_return_country(postcode)
+            if not valid_postcode:
                 self.add_error("postcode", _("Enter a valid postcode"))
+            else:
+                country, nhs_ha = valid_postcode
+                if country == "Northern Ireland":
+                    self._region = Region.NI
+                elif country == "Isle of Man":
+                    self._region = Region.IOM
+                elif country == "Channel Islands" and nhs_ha == "Jersey Health Authority":
+                    self._region = Region.JERSEY
+                elif country == "Channel Islands" and nhs_ha == "Guernsey Health Authority":
+                    self._region = Region.GUERNSEY
+                elif country == "Scotland":
+                    self._region = Region.SCOTLAND
+                elif country == "England" or country == "Wales":
+                    self._region = Region.ENGLAND_OR_WALES
+                else:
+                    self.add_error("postcode", _("This service is only available for England and Wales"))
 
         return data
 
     @property
     def region(self):
-        postcode_no_space = self.cleaned_data.get("postcode").replace(" ", "")
-        if re.search(r"^BT[0-9]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.NI
-        elif re.search(r"^IM[0-9]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.IOM
-        elif re.search(r"^JE[0-9]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.JERSEY
-        elif re.search(r"^GY[1][0]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.GUERNSEY
-        elif re.search(r"^GY[9]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.GUERNSEY
-        elif re.search(r"^GY[0-8]", postcode_no_space, flags=re.IGNORECASE):
-            return Region.GUERNSEY
-        elif postcode_no_space[:2] in SCOTTISH_PREFIXES:
-            return Region.SCOTLAND
-        else:
-            return Region.ENGLAND_OR_WALES
+        # If no region was detected, return None.
+        return getattr(self, "_region", None)
```

Comment:
should the region logic be here instead? This method is now just responsible for returning "None" which seems a little weird.